### PR TITLE
Decouple functions from topics with links

### LINKED ISSUE #560: Decouple functions from topic bindings
Functions and topics are both first class citizens in riff. However, the function resource defines references to the input and output topics implying that the function is dominant. Building towards the ideas defined in #459, where riff-cli will manage topics and function independently, we'll need to decouple the function resource from the topic bindings.

Today:
```
+-------+    input +----------+ output   +-------+
| Topic |<---------+ Function +--------->| Topic |
+-------+          +----------+          +-------+
```

Tomorrow:
```
                   +----------+
                   |          |
                   | Function |
                   |          |
                   +----------+
                        ^
                        |
+-------+               |handler        +-------+
|       |    input +----+----+ output   |       |
| Topic |<---------+ Binding +--------->| Topic |
|       |          +---------+          |       |
+-------+                               +-------+
```

In the initial pass:
- deployments/autoscaling are driven by bindings rather then functions
- use ownerRefernces to prevent a topic or function that is in use from being deleted 🎉 
- riff-cli will init/create/apply/delete a binding resource along side functions and topics (the core UX will not change until #459 is addressed)

In the future we could:
- allow a binding to support multiple input and output topics
- dynamically bind topics for a single function invocation

### PR DESCRIPTION
**Pull request by scothis**
The `input` and `output` properties of the Function CRD have moved into
a new Link CRD, the function is references with the `function`
property.

riff-cli will now scaffold and apply `*-link.yaml` along side the
existing `*-function.yaml` and `*-topics.yaml`. The UX flows of riff-cli
have not changed.

function-controller additionally watches link resources. Deployments
are created from links instead of from functions. The same function
can be linked multiple times, creating multiple deployments.

```
                 +----------+
                 |          |
                 | Function |
                 |          |
                 +----------+
                      ^
                      | function
+-------+          +--+---+          +-------+
|       |    input |      | output   |       |
| Topic |<---------+ Link +--------->| Topic |
|       |          |      |          |       |
+-------+          +------+          +-------+
```

Fixes #560

### CODE REVIEW

File: kubernetes-crds/pkg/apis/projectriff.io/v1alpha1/binding.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	meta_v1.ObjectMeta `json:"metadata"`
+	Spec               BindingSpec    `json:"spec"`
+	Status             *BindingStatus `json:"status,omitempty"`
+}
+
+// Spec (what the user wants) for a binding.
+type BindingSpec struct {
+
+	// The name of the function to bind
+	Handler string `json:"handler"`
```
Comment (ericbottard) [inline kubernetes-crds/pkg/apis/projectriff.io/v1alpha1/binding.go]: Why introduce a new term here, and not use "function"?

**Comment by scothis** [comment]
Renamed the `Handler` property to `Function` per @ericbottard's suggestion, and the `Binding` CRD to `TopicBinding` since binding is already in use by k8s.

File: riff-cli/cmd/delete.go [hunk 2]
```diff
@@ -123,20 +129,20 @@ func deleteResources(cmdArgs []string, actingClient kubectl.KubeCtl) error {
...
-	getArgs = append(getArgs, "functions.projectriff.io", opts.FunctionName, "-o", "json")
+	getArgs = append(getArgs, "topicbindings.projectriff.io", opts.FunctionName, "-o", "json")
 	json, err := queryClient.Exec(getArgs)
 	if err != nil {
-		return "", "", err
+		return "", "", "", err
 	}
 	parser := jsonpath.NewParser([]byte(json))
 
+	handler, _ := parser.StringValue(`$.spec.handler`)
```
Comment (trisberg) [inline riff-cli/cmd/delete.go]: this should be `$.spec.function` or the function resources will not get deleted, should also rename the variable I guess

**Comment by scothis** [comment]
@trisberg rebased and fixed your comments

**Comment by scothis** [comment]
I'd like for this change to get some dev usage before a release. If 0.0.7 is imminent, this PR should merge after the release.

**Comment by scothis** [comment]
@ericbottard I rebased on top of master and moved the `Windowing` config from `Function` to `Link`.

**Comment by ericbottard** [comment]
You'll need to add `links` to the list of authorized resources in `config/rbac/riff-role.yaml` and `config/rbac/riff-clusterrole.yaml`

**Comment by ericbottard** [comment]
As well as in helm-charts' `rbac.yaml`

**Comment by ericbottard** [comment]
Also, the `idleTimeoutMs` setting should be *per link* I believe

**Comment by glyn** [comment]
Fixed the remaining review issues (thanks @scothis for reviewing my changes) and merged as 77a3ba4a269f688f0f6d91de049f7bc3d3fc8bcc.

**Pull request closed without merging**